src/runtime/methodbinder.cs

```diff
@@ -505,29 +536,49 @@ static Type TryComputeClrArgumentType(Type parameterType, IntPtr argument, bool
             return clrtype;
         }
 
-        static bool MatchesArgumentCount(int argumentCount, ParameterInfo[] parameters,
+        static bool MatchesArgumentCount(int positionalArgumentCount, ParameterInfo[] parameters,
+            Dictionary<string, IntPtr> kwargDict,
             out bool paramsArray,
             out ArrayList defaultArgList)
         {
             defaultArgList = null;
             var match = false;
             paramsArray = false;
 
-            if (argumentCount == parameters.Length)
+            if (positionalArgumentCount == parameters.Length)
             {
                 match = true;
-            } else if (argumentCount < parameters.Length)
+            }
+            else if (positionalArgumentCount < parameters.Length)
             {
+                // every parameter past 'positionalArgumentCount' must have either
+                // a corresponding keyword argument or a default parameter
                 match = true;
                 defaultArgList = new ArrayList();
-                for (var v = argumentCount; v < parameters.Length; v++) {
-                    if (parameters[v].DefaultValue == DBNull.Value) {
-                        match = false;
-                    } else {
-                        defaultArgList.Add(parameters[v].DefaultValue);
+                for (var v = positionalArgumentCount; v < parameters.Length; v++)
+                {
+                    if (kwargDict.ContainsKey(parameters[v].Name))
+                    {
+                        // we have a keyword argument for this parameter,
+                        // no need to check for a default parameter, but put a null
+                        // placeholder in defaultArgList
+                        defaultArgList.Add(null);
+                    }
+                    else
+                    {
+                        // find default parameter
+                        if (parameters[v].DefaultValue == DBNull.Value)
```

Comment:
This is not quite right, I think. According to https://stackoverflow.com/questions/9977530/default-parameters-and-reflection-if-parameterinfo-isoptional-then-is-defaultva?rq=1, the CLR allows for optional values that don't supply a specific default value, in which case `default(T)` should be used.

So essentially the logic needs to be:

    if (parameters[v].IsOptional) {
        object defaultValue;
        if (parameters[v].HasDefaultValue) { // Shorter than checking for DBNull.Value in 4.5
            defaultValue = parameters[v].DefaultValue;
        }
        else {
            var type = parameters[v].ParameterType;
            defaultValue = type.IsValueType ? Activator.CreateInstance(type) : null;
        }
        defaultArgList.Add(defaultValue);
    }
    else {
        match = false;
    }

Reply:
This makes sense, I had forgotten about the use of the [OptionalAttribute] tag to designate an optional parameter.  I will integrate your suggested change and add another test case.

Reply:
One complication I ran into is that .NET 4.0 doesn't support HasDefaultValue, so I used the Attributes property to test for ParameterAttributes.HasDefault.

Also, I found [here ](https://stackoverflow.com/questions/3416216/optionalattribute-parameters-default-value)that the logic for the default value of an [Optional] parameter was slightly more complicated, so I moved it to an extension method.